pylint/checkers/unicode.py

```diff
@@ -0,0 +1,670 @@
+        while pos > 0:
+            if pos != ignore_pos:
+                result[pos] = char
+            start = pos + 1
+            pos = line.find(search_for, start)
+
+    return result
+
+
+@lru_cache(maxsize=1)
+def get_bom_sorted() -> Dict[bytes, str]:
+    """
+    Get all BOM sorted by size, biggest first
+    """
+    boms_unsorted = {
+        codecs.BOM_UTF8: "utf-8",
+        codecs.BOM_UTF16_LE: "utf-16le",
+        codecs.BOM_UTF16_BE: "utf-16be",
+        codecs.BOM_UTF32_LE: "utf-32le",
+        codecs.BOM_UTF32_BE: "utf-32be",
+    }
+    result = OrderedDict()
+    for key in sorted(boms_unsorted.keys(), key=len, reverse=True):
+        result[key] = boms_unsorted[key]
+    return result
+
+
+@lru_cache(maxsize=1000)
+def cached_encode(string: str, encoding: str) -> bytes:
+    """
+    A cached version of encode
+    """
+    return string.encode(encoding)
+
+
+def extract_codec_from_bom(first_line: bytes) -> Optional[str]:
+    """
+    Try to extract the codec (unicode only) by checking for the BOM
+
+    see https://unicode.org/faq/utf_bom.html#BOM
+
+    Default to UTF-8 if no BOM is found
+
+    Args:
+        first_line: the first line of a file
+
+    Returns:
+        a codec name
+    """
+    for bom, codec in get_bom_sorted().items():
+        if first_line.startswith(bom):
+            return codec
+
+    return None
+
+
+class InvalidUnicodeModeChecker(BaseChecker):
+    """
+    Check that UTF-16 and UTF-32 are not used to encode Python Files.
+
+    Currently Python supports only UTF-8. See
+    https://stackoverflow.com/questions/69897842/ and https://bugs.python.org/issue1503789
+    for background.
+    This simplifies the BiDirectional Unicode check below as we only need to check
+    for UTF-8 bytes.
+
+    As this is a security related check we want to ensure that even once Python
+    support UTF-16 and UTF-32 it is not possible for an attacker to bypass our checks
+    by simply using a new encoding.
+    """
+
+    __implements__ = interfaces.IRawChecker
+
+    msgs = {
+        "E1800": (
+            "Use only UTF-8, not UTF-16 or UTF-32!",
+            "invalid-unicode-codec",
+            (
+                "UTF-16 / UTF-32 is currently not working properly with Python, "
+                "see https://stackoverflow.com/questions/69897842/"
+            ),
+        )
+    }
+    name = "invalid_unicode_encoding_checker"
+
+    @staticmethod
+    def is_invalid_codec(codec: Optional[str]) -> bool:
+        if codec is None:
+            return False
+        codec = codec.lower().replace("-", "").replace("_", "")
+        return codec.startswith("utf16") or codec.startswith("utf32")
+
+    def process_module(self, node: nodes.Module) -> None:
+        """
+        Perform the the actual checking by checking module stream
+
+        the module's content is accessible via node.stream() function
+        """
+        codec: Optional[str]
+        lines: Sequence[bytes]
```

Comment:
Shouldn't this be optional for if the `try` fails?

Reply:
If the try fails ``lines`` is not set at all.
I know this looks a bit more like *C* instead of Python, but I don't know a good syntax to define the types if two results are given.

``codec`` is Optional as the result of ``extract_codec_from_bom`` can be None (=Nothing found), not because the try failed.

Please resolve if you find this correct :-)